pkg/autoscaler/scaling/autoscaler.go

```diff
@@ -245,8 +244,43 @@ func (a *Autoscaler) Scale(ctx context.Context, now time.Time) (desiredPodCount
 	return desiredPodCount, int32(excessBCF), true
 }
 
+func clampInclusiveRange(valueToClamp, lowerBound, upperBound float64) float64 {
+	return math.Min(math.Max(valueToClamp, lowerBound), upperBound)
+}
+
+// Query functions
+func (a *Autoscaler) currentlyStable() bool {
+	return a.panicTime.IsZero()
+}
+
+func (a *Autoscaler) currentlyPanicking() bool {
+	return !a.currentlyStable()
+}
+
+func (a *Autoscaler) stableWindowHasElapsed(spec *DeciderSpec, now time.Time) bool {
+	return a.panicTime.Add(spec.StableWindow).Before(now)
+}
+
 func (a *Autoscaler) currentSpecAndPC() (*DeciderSpec, resources.EndpointsCounter) {
 	a.specMux.RLock()
 	defer a.specMux.RUnlock()
 	return a.deciderSpec, a.podCounter
 }
+
+// Panic functions
+func (a *Autoscaler) updatePanic(panicTime time.Time, maxPanicPods int32) {
+	a.panicTime = panicTime
+	a.maxPanicPods = maxPanicPods
+}
+
+func (a *Autoscaler) startPanicking(now time.Time, logger *zap.SugaredLogger) {
+	logger.Info("PANICKING.")
+	a.updatePanic(now, 0)
```

Comment:
shouldn't we pass current scale as well to set `maxPanicPods`. it will be overwritten below `startPanicking` call, but still would  be more correct.

Reply:
I think I assumed that the previous code relied on the fact that `maxPanicPods` is set to zero when a panic ends. 

Tracing the variable back, it looks like setting it zero when a panic begins would work for new Revisions, but not for controller restarts. 

I'll update with a test case.